### PR TITLE
Frontend V2 - Add/Remove Courses From Semesters

### PR DESCRIPTION
**Pull request by aryanshah701**
# Description

Reminder: This PR is "stacked" on top of v2-plan-handle-errors. This diff will only show changes specific to adding/removing a course from a semester.

https://user-images.githubusercontent.com/30478978/187541665-f9603935-07ea-44c0-a7fd-b39840503056.mov

Adds support to add and remove courses from semester in a plan.

An overview of all the functionality completed and remaining for the plan component(note, there may be more stuff remaining but this is what i can think of for now).

- [x] fetch plans and populate dnd ids
- [x] basic plan component with dnd and posting to API
- [x] styling
- [x] collapsing years
- [x] choosing which plan to display
- [x] handle home page errors: previous PR
- [x] **adding a course using the add course modal: current PR**
- [ ] adding a Plan
- [ ] choosing primary plan
  - [ ] fix backend bug of allowing users to choose a primary plan that doesn't belong to this(massive security issue) 
- [ ] validation and errors
- [ ] display and edit semester type (add Co-op block too)
- [ ] edit plans
- [ ] add class validator decorators to schedule type so that we enforce at least some shape of the schedule when posting
- [ ] coreq, prereq errors
- [ ] figure out how to show "preview" of course in the position it was when it is being dragged around
- [ ] batch/throttle plan updates

## Type of change

Please delete options that are not relevant.

- [x] New feature (non-breaking change which adds functionality)

# How Has This Been Tested?

Manually. Check video.

# Checklist:

- [x] I have performed a self-review of my own code
- [x] I have commented my code where needed 
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

**Review by rael346 — APPROVED** [review]
Awesome PR! Most of the comments are just clarification, but there is a big one regarding the current plan state.

File: packages/api-client/index.ts [hunk 1]
```diff
@@ -13,6 +13,7 @@ import {
   UpdateStudentDto,
   UpdateStudentResponse,
   OnboardStudentDto,
+  ScheduleCourse2,
```
[the remark this answers is not in the document]
Comment (aryanshah701) [reply packages/api-client/index.ts:16]: ahh yes, will remove

File: packages/api-client/index.ts [hunk 2]
```diff
@@ -123,7 +124,8 @@ class SearchAPIClient {
 
     const courseData = await res.data.data;
     if (courseData?.class?.latestOccurrence) {
-      const course: ScheduleCourse = courseData.class.latestOccurrence;
+      const course: ScheduleCourse2<unknown> =
```
[the remark this answers is not in the document]
Comment (aryanshah701) [reply packages/api-client/index.ts:127]: you're right this should be null!

File: packages/frontend-v2/components/AddCourseModal/SearchCoursesInput.tsx [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+
+  return (
+    <InputGroup>
+      <InputLeftElement pointerEvents="none">
+        <Search2Icon color="primary.blue.light.main" />
+      </InputLeftElement>
+      <Input
+        value={searchTerm}
+        onChange={(e) => {
+          setSearchTerm(e.target.value.trim());
```
[the remark this answers is not in the document]
Comment (aryanshah701) [reply packages/frontend-v2/components/AddCourseModal/SearchCoursesInput.tsx:37]: ooh we actually do allow the user to press spacebar, we just trim any spaces at the end and beginning of the string

File: packages/frontend-v2/utils/plan/addClassesToTerm.ts [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+import produce from "immer";
+import { prepareClassesForDnd } from ".";
+import { logger } from "../logger";
+import { flattenScheduleToTerms } from "./updatePlanOnDragEnd";
+
+/**
+ * Add the given classes to the given term in the plan.
+ *
+ * @returns The updated plan. The original plan isn't touched and remains unchanged.
+ */
```
[the remark this answers is not in the document]
Comment (aryanshah701) [reply packages/frontend-v2/utils/plan/addClassesToTerm.ts:11]: decoupling data processing from api communication :))

File: packages/frontend-v2/utils/plan/updatePlanForStudent.ts [hunk 5]
```diff
@@ -1,4 +1,5 @@
 import { StudentModel, PlanModel } from "@graduate/common";
+import produce, { produceWithPatches } from "immer";
 
 /**
  * Creates a copy of the given student and replaces one of their plans with
```
[the remark this answers is not in the document]
Comment (aryanshah701) [reply packages/frontend-v2/utils/plan/updatePlanForStudent.ts:5]: mutating state directly gets weird, react is very functional so it's always safer to create something new than mutate

File: packages/frontend-v2/components/Plan/Plan.tsx [hunk 6]
```diff
@@ -46,6 +85,11 @@ export const Plan: React.FC<PlanProps> = ({ plan }) => {
               scheduleYear={scheduleYear}
               isExpanded={isExpanded}
               toggleExpanded={() => toggleExpanded(scheduleYear)}
+              isCourseInCurrPlan={(course: ScheduleCourse2<unknown>) =>
+                isCourseInPlan(course, plan)
```
[the remark this answers is not in the document]
Comment (aryanshah701) [reply packages/frontend-v2/components/Plan/Plan.tsx:89]: interesting idea, so we display the last plan the user was working on when the visit the app. I like it, lets bring this up with the designers!

**Pull request merged**